ocaml/typing/solver.ml

```diff
@@ -0,0 +1,779 @@
+open Solver_intf
+
+module Magic_allow_disallow (X : Allow_disallow) :
+  Allow_disallow with type ('a, 'b, 'd) t := ('a, 'b, 'd) X.t = struct
+  let disallow_right :
+      type a b l r. (a, b, l * r) X.t -> (a, b, l * disallowed) X.t =
+    Obj.magic
+
+  let disallow_left :
+      type a b l r. (a, b, l * r) X.t -> (a, b, disallowed * r) X.t =
+    Obj.magic
+
+  let allow_right : type a b l r. (a, b, l * allowed) X.t -> (a, b, l * r) X.t =
+    Obj.magic
+
+  let allow_left : type a b l r. (a, b, allowed * r) X.t -> (a, b, l * r) X.t =
+    Obj.magic
+end
+
+(** Error returned by failed [submode a b]. [left] will be the lowest mode [a]
+   can be, and [right] will be the highest mode [b] can be. And [left <= right]
+   will be false, which is why the submode failed. *)
+type 'a error =
+  { left : 'a;
+    right : 'a
+  }
+
+let rec find_error f = function
+  | [] -> Ok ()
+  | x :: rest -> (
+    match f x with Ok () -> find_error f rest | Error _ as e -> e)
+
+module VarSet = Set.Make (Int)
+
+module Solver_mono (C : Lattices_mono) = struct
+  type 'a var =
+    { mutable upper : 'a;
+      mutable lower : 'a;
+      (* TODO: consider using hashset for quicker deduplication *)
+      mutable vlower : 'a lmorphvar list
+    }
+
+  and 'b lmorphvar = ('b, left_only) morphvar
+
+  and ('b, 'd) morphvar =
+    | Amorphvar : 'a var * ('a, 'b, 'd) C.morph -> ('b, 'd) morphvar
+
+  type change =
+    | Cupper : 'a var * 'a -> change
+    | Clower : 'a var * 'a -> change
+    | Cvlower : 'a var * 'a lmorphvar list -> change
+
+  type changes = change list
+
+  let undo_change = function
+    | Cupper (v, upper) -> v.upper <- upper
+    | Clower (v, lower) -> v.lower <- lower
+    | Cvlower (v, vlower) -> v.vlower <- vlower
+
+  let undo_changes l = List.iter undo_change l
+
+  (* To be filled in *)
+  let append_changes : (changes ref -> unit) ref = ref (fun _ -> assert false)
+
+  type ('a, 'd) mode =
+    | Amode : 'a -> ('a, 'l * 'r) mode
+    | Amodevar : ('a, 'd) morphvar -> ('a, 'd) mode
+    | Amodejoin :
+        'a * ('a, 'l * disallowed) morphvar list
+        -> ('a, 'l * disallowed) mode
+    | Amodemeet :
+        'a * ('a, disallowed * 'r) morphvar list
+        -> ('a, disallowed * 'r) mode
+
+  let address_of : 'a var -> int = Obj.magic
+
+  let rec print_var : type a. ?traversed:VarSet.t -> a C.obj -> _ -> a var -> _
```

Comment:
There can be valid cycles. I will elaborate in comments.

Reply:
I haven't spotted these comments (though perhaps I've looked in the wrong spot).

Reply:
Apologies - just added.

Reply:
Thanks for this example. But it's pretty abstract, defining a new lattice and morphism, neither of which exist as described. (Though maybe composition of two morphisms on the Regionality lattice would be the same?) I guess my real question is: does this ever actually happen when processing OCaml code? Even if not, this is a debugging function, and so might usefully retain its loop checking (because perhaps a bug causes a loop). As is, the comment addresses, but does not completely answer the question of "why is this code here?".

Reply:
yes composition of two morphisms wrt locality/regionality would give such a morphism (though, I would try to avoid mentioning `mode.ml` in `solver.ml`). 

I vaguely remember seeing mode cycles when type checking. At the least, I don't have enough reasoning to rule this out.

Reply:
As discussed, cycles will be investigated in the future. Resolving.